mesa/experimental/cell_space/cell_agent.py

```diff
@@ -1,37 +1,83 @@
+
+    def move_relative(self, directions: tuple[int, ...], distance: int = 1): ...
+
+
+class CellAgent:
+    """Cell Agent is an Agent class that adds behavior for moving in discrete spaces
 
     Attributes:
-        unique_id (int): A unique identifier for this agent.
-        model (Model): The model instance to which the agent belongs
-        pos: (Position | None): The position of the agent in the space
-        cell: (Cell | None): the cell which the agent occupies
+        space (DiscreteSpace): the discrete space the agent is in
+        cell (Cell): the cell the agent is in
     """
 
-    def __init__(self, model: Model) -> None:
-        """
-        Create a new agent.
+    def __init__(
+        self,
+        space: DiscreteSpace[Cell],
+        cell: Cell | None = None,
+        *args: tuple[Any],
+        **kwargs: dict[str, Any],
+    ):
+        super().__init__(*args, **kwargs)
+        self.space = space
+        self.cell = cell
+        if cell is not None:
+            cell.add_agent(self)
 
-        Args:
-            unique_id (int): A unique identifier for this agent.
-            model (Model): The model instance in which the agent exists.
-        """
-        super().__init__(model)
-        self.cell: Cell | None = None
+    @property
+    def coordinate(self) -> tuple[int, ...]:
+        return self.cell.coordinate if self.cell else ()
 
-    def move_to(self, cell) -> None:
+    def move_to(self, cell: Cell) -> None:
         if self.cell is not None:
             self.cell.remove_agent(self)
         self.cell = cell
         cell.add_agent(self)
+
+    def move_relative(self, directions: tuple[int, ...], distance: int = 1):
+        new_position = tuple(
+            self.cell.coordinate[i] + directions[i] * distance
+            for i in range(len(directions))
+            if self.cell
+        )
+        new_cell = self.space[new_position]
+        self.move_to(new_cell)
+
+
+class Grid2DMovingAgent:
+    def move(self: DiscreteSpaceAgent[Cell], direction: str, distance: int = 1):
+        match direction:
+            case "N" | "North" | "Up":
+                self.move_relative((-1, 0), distance)
+            case "S" | "South" | "Down":
+                self.move_relative((1, 0), distance)
+            case "E" | "East" | "Right":
+                self.move_relative((0, 1), distance)
+            case "W" | "West" | "Left":
+                self.move_relative((0, -1), distance)
+            case "NE" | "NorthEast" | "UpRight":
+                self.move_relative((-1, 1), distance)
+            case "NW" | "NorthWest" | "UpLeft":
+                self.move_relative((-1, -1), distance)
+            case "SE" | "SouthEast" | "DownRight":
+                self.move_relative((1, 1), distance)
+            case "SW" | "SouthWest" | "DownLeft":
+                self.move_relative((1, -1), distance)
+            case _:
+                raise ValueError(f"Invalid direction: {direction}")
```

Comment:
1. Great use of pattern matching!
2. Should we also allow lower case? If so, we can do `direction = direction.lower()` and then match against the lower case strings for example.
3. This now allows "queen" (chess) moves. Do we also want to allow knights and maybe even more? (south-south west, etc.)
4. Do we want to allow a certain amount of degrees? 0 is up, 90 is east, 135 is south east, etc.
5. Do we want to link this to the type of grid in any way? Because some grids (currently) have neighbours cells that are in different directions. Maybe add a filter "only_if_neighbour" or something better named than that.

I think none of these are blocking for this PR, just a few thoughts.

Reply:
1. Thanks! 
2. Yes, they should also work with lowercase directions I think. 
3. This is supposed to be used on a regular grid, where each cell has 8 neighbboring cells it touches. So these are the base movements, everything else can be composed of these. This is of course a bit arbitrary, technically 4 base movements are sufficient (linked to von Neumann vs Moore). But I think its a convenient set
4. I think degrees deserve a separate implementation. Actually my first implementation was heading more towards degrees, where each agent has a "heading" and movement is implemented in terms of forward/backward/turnLeft/turnRight. I think this kind of movement would also be nice!
5. Well the class is specifically called Grid2DMovingAgent to address this. It only works with 2DGrids, not necessarily with higher dimensions or other subclasses of `DiscreteSpace`